spring-kafka/src/test/java/org/springframework/kafka/annotation/KafkaListenerTests.java

```diff
@@ -0,0 +1,234 @@
+
+		@Autowired
+		EmbeddedKafkaBroker broker;
+
+		@Bean
+		public KafkaListenerContainerFactory<?> kafkaListenerContainerFactory() {
+			ConcurrentKafkaListenerContainerFactory<Integer, String> factory =
+					new ConcurrentKafkaListenerContainerFactory<>();
+			factory.setConsumerFactory(consumerFactory());
+			return factory;
+		}
+
+		@Bean
+		public DefaultKafkaConsumerFactory<Integer, String> consumerFactory() {
+			return new DefaultKafkaConsumerFactory<>(consumerConfigs());
+		}
+
+		@Bean
+		public Map<String, Object> consumerConfigs() {
+			Map<String, Object> consumerProps = KafkaTestUtils.consumerProps(
+					this.broker.getBrokersAsString(),
+					"myDefaultListenerGroup",
+					"false");
+			return consumerProps;
+		}
+
+		@Bean
+		public KafkaTemplate<Integer, String> template() {
+			return new KafkaTemplate<>(producerFactory());
+		}
+
+		@Bean
+		public ProducerFactory<Integer, String> producerFactory() {
+			return new DefaultKafkaProducerFactory<>(producerConfigs());
+		}
+
+		@Bean
+		public Map<String, Object> producerConfigs() {
+			return KafkaTestUtils.producerProps(this.broker.getBrokersAsString());
+		}
+
+		@Component
+		@KafkaListener(id = "classAndMethodLevelListener", topics = "default-listener.tests")
+		public static class ClassAndMethodLevelListener {
+
+			final CountDownLatch latch = new CountDownLatch(1);
+
+			@KafkaHandler
+			public void listen(String in) {
+				this.latch.countDown();
+			}
+		}
+
+		@Component
+		@KafkaListener(id = "classLevelListenerWithSinglePublicMethod", topics = TEST_TOPIC)
+		public static class ClassLevelListenerWithSinglePublicMethod {
+
+			final CountDownLatch latch = new CountDownLatch(1);
+
+			public void listen(String in) {
+				this.latch.countDown();
+			}
+		}
+
+		@Component
+		@KafkaListener(id = "otherClassLevelListenerWithSinglePublicMethod", topics = TEST_TOPIC)
+		public static class OtherClassLevelListenerWithSinglePublicMethod {
+
+			final CountDownLatch latch = new CountDownLatch(1);
+
+			public void listen(String in) {
+				this.latch.countDown();
+			}
+		}
+
+		@Component
+		@KafkaListener(id = "classLevelListenerWithDoublePublicMethod", topics = TEST_TOPIC)
+		public static class ClassLevelListenerWithDoublePublicMethod {
+
+			public final static int INIT_LATCH_COUNT = 2;
+
+			final CountDownLatch latch = new CountDownLatch(INIT_LATCH_COUNT);
+
+			public void listen1(String in) {
+				this.latch.countDown();
+			}
+
+			public void listen2(String in) {
+				this.latch.countDown();
+			}
+		}
+
+		@Component
+		@KafkaListener(id = "classLevelListenerWithSinglePublicMethodAndPrivateMethod", topics = TEST_TOPIC)
+		public static class ClassLevelListenerWithSinglePublicMethodAndPrivateMethod {
+
+			final CountDownLatch latch = new CountDownLatch(1);
+
+			public void listen1(String in) {
+				this.latch.countDown();
+			}
+
+			private void listen2(String in) {
+				this.latch.countDown();
+			}
+
+			private void listen3(String in) {
```

Comment:
This variant does not make sense since we don't take into account private methods.